### PR TITLE
Added token authentication test.

### PR DESCRIPTION
**Pull request by Arthasasyan**

### CODE REVIEW

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenRepository.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                    .queryParam("user", user)
+                    .queryParam("username",user)
+                    .queryParam("method", method)
+                    .queryParam("urls", url)
+                    .queryParam("expirationDate")
+                    .queryParam("Addresses")
+                    .queryParam("create", "Issue+Token")
+                    .request()
+                    .method("POST"));
+        String[] responseUrl = response.getHeaderString("Location").split(";")[0].split("/");
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/tokens/TokenRepository.java]: Use `StringUtils.substringAfter(";")`

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenRepository.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+        String requestString ="/admin/users/tokens/new";
+        Response response = executeRootRequest(webTarget -> webTarget.path(requestString)
+                    .queryParam("user", user)
+                    .queryParam("username",user)
+                    .queryParam("method", method)
+                    .queryParam("urls", url)
+                    .queryParam("expirationDate")
+                    .queryParam("Addresses")
+                    .queryParam("create", "Issue+Token")
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/tokens/TokenRepository.java]: username, method, urls are enough

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenRepository.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import com.axibase.tsd.api.method.BaseMethod;
+
+import org.testng.annotations.Test;
+
+import io.qameta.allure.Issue;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+@Issue("6052")
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/tokens/TokenRepository.java]: `@Issue` annotation is to be set only on test methods.

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                                        .queryParam("create", "true")
+                                        .queryParam("current", "false")
+                                        .request()
+                                        .method("POST"));
+    }
+
+    private void deleteUser(String username)
+    {
+        String path ="/admin/users/edit.xhtml";
+        executeRootRequest(webTarget -> webTarget.path(path)
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java]: userBean.username and delete are enough

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        return executeApiRequest(webTarget -> webTarget.path(path)
+        .request()
+        .header("Authorization: Bearer", token)
+        .method(method, Entity.json("entity")));
+    }
+
+    private void createUser(String username, String password)
+    {
+        String path ="/admin/users/edit.xhtml";
+        executeRootRequest(webTarget -> webTarget.path(path)
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java]: required parameters: enabled, userBean.username, userBean.password, repeatPassword, userBean.userRoles, save, create

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            }
+        }
+        deleteUser(username);
+    }
+
+    private Response executeMethodWithoutEntity(String token, String path, String method)
+    {
+        return executeApiRequest(webTarget -> webTarget.path(path)
+        .request()
+        .header("Authorization: Bearer", token)
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java]: Use HttpHeaders constants. `Bearer` must be in value

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        {"/entity-groups/group/entities/add", "POST"},
+        {"/entity-groups/group/entities/set", "POST"},
+        {"/entity-groups/group/entities/delete", "POST"},
+        {"/search", "GET"},
+        {"/version", "GET"},
+    };
+
+    @Issue("6052")
+    @Test
+    public void tokenAccessTest() throws Exception {
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java]: Parameterize test

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+import com.axibase.tsd.api.method.BaseMethod;
+
+import com.axibase.tsd.api.Config;
+
+import org.testng.annotations.Test;
+
+import io.qameta.allure.Issue;
+
+
+@Issue("6052")
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java]: Annotations must be specified on test methods

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        .request()
+        .header("Authorization: Bearer", token)
+        .method(method));
+    }
+    private Response executeMethodWithEntity(String token, String path, String method)
+    {
+        return executeApiRequest(webTarget -> webTarget.path(path)
+        .request()
+        .header("Authorization: Bearer", token)
+        .method(method, Entity.json("entity")));
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java]: ?

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java (hunk 7 above)
Comment (raipc) [reply src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java]: Tests must perform a valid request and check that response is valid too. It's not necessary to test all endpoints.
But we need test for different users: admin and non-admin with permissions to some entities. Check that only specified entities are available for that user.

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+@Issue("6052")
+@Test
+public class TokenAuthenticationTest extends BaseMethod {
+
+    private static final String[][] availablePaths = new String[][]{
+        // Data API
+        {"/series/query", "POST"},
+        {"/series/insert", "POST"},
+        {"/series/csv/entity", "POST"},
+        {"/series/format/entity/metric", "GET"},
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/tokens/TokenAuthenticationTest.java]: {format} is a placeholder, may be `json` or `csv`

**Comment by Arthasasyan** [comment]
Token Access tests are failed because token insertion was incorrect:

```java
com.axibase.tsd.api.method.tokens.TokenAccessTest.tokenAccessToItsUrlTest(com.axibase.tsd.api.method.tokens.TokenAccessTest)
  Run 1: PASS
  Run 2: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /series/query and method POST failed to authorise user apitokenuser token: mfhU4RlqppPBxbJMmRfrOMR7C0626ht0H5Sr response: {"error":"code 03"}
  Run 3: PASS
  Run 4: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /series/insert and method POST failed to authorise user apitokenuser token: C4hbya2Br1MfznmTJouePMyWX4whtf3j64ra response: {"error":"code 03"}
  Run 5: PASS
  Run 6: PASS
  Run 7: PASS
  Run 8: PASS
  Run 9: PASS
  Run 10: PASS
  Run 11: PASS
  Run 12: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /series/query and method POST failed to authorise user apitokenuser token: mfhU4RlqppPBxbJMmRfrOMR7C0626ht0H5Sr response: {"error":"code 03"}
  Run 13: PASS
  Run 14: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /series/delete and method POST failed to authorise user apitokenuser token: 3G311VPJmKaePSBkymF6S430VftmBNIqapf0 response: {"error":"code 03"}
  Run 15: PASS
  Run 16: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /properties/query and method POST failed to authorise user apitokenuser token: quhdSG3XGw2lG6DP9ZeEIm0whsbQKiy0jU5q response: {"error":"code 03"}
  Run 17: PASS
  Run 18: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /properties/insert and method POST failed to authorise user apitokenuser token: X0Hi0IUWsx2GPFiDuf4ZEDHrUHulOW2X4UfP response: {"error":"code 03"}
  Run 19: PASS
  Run 20: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /properties/delete and method POST failed to authorise user apitokenuser token: Qc6MP7ivDnsAEGqsvj0A1jSTjGB2EyHPVB55 response: {"error":"code 03"}
  Run 21: PASS
  Run 22: PASS
  Run 23: PASS
  Run 24: PASS
  Run 25: PASS
  Run 26: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /messages/query and method POST failed to authorise user apitokenuser token: cLG972NF2tdNm0p0sR1bMqJHObSc85PV4etG response: {"error":"code 03"}
  Run 27: PASS
  Run 28: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /messages/insert and method POST failed to authorise user apitokenuser token: 5vErEdQuDBSyfCDTivZK8sBSNGupADOMbABw response: {"error":"code 03"}
  Run 29: PASS
  Run 30: PASS
  Run 31: PASS
  Run 32: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /messages/stats/query and method POST failed to authorise user apitokenuser token: P6ICgpe8xPWHB0gdODPvP4koTWNc2D4oYiQe response: {"error":"code 03"}
  Run 33: PASS
  Run 34: PASS
  Run 35: PASS
  Run 36: PASS
  Run 37: PASS
  Run 38: PASS
  Run 39: PASS
  Run 40: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /alerts/query and method POST failed to authorise user apitokenuser token: TXERaMye55hSUMyAdpRw3k7rgjn8V1NBJJH6 response: {"error":"code 03"}
  Run 41: PASS
  Run 42: PASS
  Run 43: PASS
  Run 44: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /alerts/delete and method POST failed to authorise user apitokenuser token: 3viok8OQwk09yU1Z48BAtbdd0mCA7TuUD98q response: {"error":"code 03"}
  Run 45: PASS
  Run 46: PASS
  Run 47: PASS
  Run 48: TokenAccessTest.tokenAccessToItsUrlTest:170 Token for path /alerts/history/query and method POST failed to authorise user apitokenuser token: GfSDnzlCT5GAHfYJShVr5wTi6NoQtjK8WL6H response: {"error":"code 03"}
```

File: src/main/java/com/axibase/tsd/api/method/checks/AlertsCheck.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    private Collection<AlertQuery> alerts;
+
+    public AlertsCheck(Collection<AlertQuery> alerts) {
+        this.alerts = alerts;
+    }
+
+    @Override
+    public boolean isChecked() {
+        try {
+            if(Response.Status.Family.CLIENT_ERROR.equals(Util.responseFamily(AlertMethod.queryAlerts(alerts)))) {
```
Comment (raipc) [inline src/main/java/com/axibase/tsd/api/method/checks/AlertsCheck.java]: Reformat code in all modified classes: Code > Reformat Code in IDEA

File: src/main/java/com/axibase/tsd/api/method/BaseMethod.java [hunk 12]
```diff
@@ -63,6 +65,11 @@
                 .property(ClientProperties.READ_TIMEOUT, DEFAULT_CONNECT_TIMEOUT)
                 .property(ClientProperties.CONNECT_TIMEOUT, DEFAULT_CONNECT_TIMEOUT)
                 .property(ClientProperties.REQUEST_ENTITY_PROCESSING, RequestEntityProcessing.BUFFERED);
+        ClientConfig tokenConfig = new ClientConfig();
+        tokenConfig.connectorProvider(new ApacheConnectorProvider());
```
Comment (raipc) [inline src/main/java/com/axibase/tsd/api/method/BaseMethod.java]: Chain method calls

File: src/main/java/com/axibase/tsd/api/method/BaseMethod.java [hunk 13]
```diff
@@ -43,6 +43,8 @@
 
     private static final GenericObjectPool<HttpClient> apiTargetPool;
     private static final GenericObjectPool<HttpClient> rootTargetPool;
+    private static final GenericObjectPool<HttpClient> tokenTargetPool;
```
Comment (raipc) [inline src/main/java/com/axibase/tsd/api/method/BaseMethod.java]: Are separate pools actually required? Why not just use root pool?

**Comment by Arthasasyan** [comment]
It seems like `NotPassedCheck` does not work as expected. It's main use case is to check that data was deleted or was not inserted in ATSD. Because ATSD needs it's time to insert data, I decided to use Checks. However, at the moment `NotPassedCheck.isChecked` will return true after the first iteration, because another `AbstractCheck` will return false if it failed to find data for the first time. 
The only solution I found is the following:
```java
package com.axibase.tsd.api.method.checks;


import com.axibase.tsd.api.Checker;
import com.axibase.tsd.api.util.NotCheckedException;


public class NotPassedCheck{

    public static void assertNotPassed(String message, AbstractCheck check) {
        try {
            Checker.check(check);
        } catch (NotCheckedException nce) {
            return;
        }
        throw new AssertionError(message);
    }
}
```

However, because of current UPPER_BOUND_FOR_CHECK value that equals to 100 seconds, this can increase test time dramatically. @raipc do you have any suggestions?

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenWorkTest.java [hunk 14]
```diff
@@ -0,0 +1,764 @@
...
+                .request()
+                .method(HttpMethod.POST))
+                .bufferEntity();
+    }
+
+
+    @Data
+    @AllArgsConstructor
+    private static final class ResponsePair {
+        private final Response responseWithApi;
```
[the remark this answers is not in the document]
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/tokens/TokenWorkTest.java]: Resolved

**Comment by Arthasasyan** [comment]
@raipc should bin folder be included in .gitignore?

**Comment by Arthasasyan** [comment]
Opened #551 due to property errors

File: src/test/java/com/axibase/tsd/api/util/TestNameGenerator.java [hunk 15]
```diff
@@ -114,7 +114,8 @@ private String camelToLisp(String camelCaseName) {
         ENTITY_GROUP("entity-group"),
         MESSAGE("message"),
         PROPERTY("property"),
-        PROPERTY_TYPE("property-type");
+        PROPERTY_TYPE("property-type"),
+        REPLACEMENT_TABLE("replacament-table");
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/util/TestNameGenerator.java]: replacement

Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/util/TestNameGenerator.java]: Fixed with the next commit

**Comment by raipc** [comment]
> Opened #551 due to property errors
> Pending till it is resolved

Model classes are not forced to return JSON representation in toString method. Use serializer to get JSON representation, as it's done in other use cases. Never build JSON manually.

File: src/test/java/com/axibase/tsd/api/method/tokens/TokenWorkTest.java [hunk 16]
```diff
@@ -0,0 +1,655 @@
...
+                {ADMIN_NAME},
+                {USER_NAME}
+        };
+    }
+
+    @BeforeClass
+    private void createUser() {
+        String path = "/admin/users/edit.xhtml";
+
+        executeRootRequest(webTarget -> webTarget.path(path)
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/tokens/TokenWorkTest.java]: Move to a separate class

File: src/main/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java [hunk 17]
```diff
@@ -48,11 +50,45 @@ public static void createCheck(ReplacementTable table) {
     private static Response getReplacementTablesResponse(String replacementTableName) {
         Response response = executeApiRequest(webTarget ->
                 resolveTable(webTarget, replacementTableName)
-                .request().get());
+                        .request().get());
         response.bufferEntity();
         return response;
     }
 
+    public static boolean replacementTableExist(ReplacementTable replacementTable) throws NotCheckedException {
```
Comment (raipc) [inline src/main/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java]: Why copy?

**Comment by raipc** [comment]
The main reason why this PR is being reviewed for so long is its size. There is a lot of copying from other tests that creates a bunch of questions why this is done. In most cases you already have all methods implemented, the only thing which needs to be done is to set appropriate Authorization header. Consider to move methods to work with specific objects to classes like MetricMethod, EntityMethod, etc. Use method delegation, not copying of method body.
Test method must create no more than two requests: one for data modification, one for check. If it doesn't, move it to a separate class and split.
Rename new *Check classes. Abstract ObjectCheck must not contain error message "object cannot be deleted".

File: src/main/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java (hunk 17 above)
Comment (Arthasasyan) [reply src/main/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java]: This method is not a total copy. It checks that received replacement equals to the given: 

https://github.com/axibase/atsd-api-test/blob/c6a22a428f7cd3737ad9c98acf8eedfcb2ca79d2/src/main/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java#L77

However, the current implementation is not fully correct - it should throw a `NotCheckedException` if replacement table does not equal to the given one. I'll fix it with the next commit.

Comment (raipc) [reply src/main/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java]: > This method is not a total copy. It checks that received replacement equals to the given:

If 28 lines of code out of 32 are exactly the same I call it a copy. The methods must be refactored.

Comment (Arthasasyan) [reply src/main/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java]: I suggest changing it to the following:

```java
private static ReplacementTable getReplacementTableAndCheckName(final String replacementTableName) throws NotCheckedException {
        String replacementTableNameLowerCase = replacementTableName.replace(" ", "_").toLowerCase();
        final Response response = getReplacementTablesResponse(replacementTableNameLowerCase);
        if (Response.Status.Family.SUCCESSFUL != Util.responseFamily(response)) {
            if (response.getStatus() == NOT_FOUND.getStatusCode()) {
                return null;
            }
            String message = "Fail to execute replacement table query: " + response.getStatusInfo();
            log.error(message);
            throw new NotCheckedException(message);
        }

        try {
            ReplacementTable replacementTable = response.readEntity(ReplacementTable.class);
            if (!StringUtils.equalsIgnoreCase(replacementTable.getName(), replacementTableName)) {
                String message = "ReplacementTable API returned an entry we weren't asking for.";
                log.error(message);
                throw new NotCheckedException(message);
            }

            return replacementTable;

        } catch (ProcessingException err) {
            NotCheckedException exception = new NotCheckedException("Could not parse Replacement Table from JSON: " + err.getMessage());
            exception.addSuppressed(err);
            log.error(exception.getMessage());
            throw exception;
        }
    }

    public static boolean replacementTableExist(ReplacementTable replacementTable) throws NotCheckedException {
        ReplacementTable receivedReplacementTable = getReplacementTableAndCheckName(replacementTable.getName());

        if (receivedReplacementTable == null) {
            return false;
        }
        for (Map.Entry<String, String> entry : receivedReplacementTable.getKeys().entrySet()) {
            if (!replacementTable.getKeys().get(entry.getKey()).equals(entry.getValue())) {
                String message = "ReplacementTable with the name " + replacementTable.getName() + " exists, but does not equal to the given one.";
                throw new NotCheckedException(message);
            }
        }
        return true;
    }

    public static boolean replacementTableExist(String replacementTableName) throws NotCheckedException {
        ReplacementTable receivedReplacementTable = getReplacementTableAndCheckName(replacementTableName);

        if (receivedReplacementTable == null) {
            return false;
        }
        return true;
    }
```

Comment (raipc) [reply src/main/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java]: split into several methods:
`findReplacementTable(String name)`
`checkReplacementTableName(ReplacementTable table, String expectedName)`
`checkReplacementTableContent(ReplacementTable table, Map<String, String> expectedContent)`

Comment (raipc) [reply src/main/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java]: remove `exception.addSuppressed(err);`

Comment (Arthasasyan) [reply src/main/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java]:
```java
private static ReplacementTable findReplacementTable(final String replacementTableName) throws NotCheckedException {
        String replacementTableNameLowerCase = replacementTableName.replace(" ", "_").toLowerCase();
        final Response response = getReplacementTablesResponse(replacementTableNameLowerCase);
        if (Response.Status.Family.SUCCESSFUL != Util.responseFamily(response)) {
            if (response.getStatus() == NOT_FOUND.getStatusCode()) {
                return null;
            }
            String message = "Fail to execute replacement table query: " + response.getStatusInfo();
            log.error(message);
            throw new NotCheckedException(message);
        }

        try {
            ReplacementTable replacementTable = response.readEntity(ReplacementTable.class);

            return replacementTable;

        } catch (ProcessingException err) {
            NotCheckedException exception = new NotCheckedException("Could not parse Replacement Table from JSON: " + err.getMessage());
            log.error(exception.getMessage());
            throw exception;
        }
    }

    private static void checkReplacementTableName(ReplacementTable replacementTable, String expectedName) throws NotCheckedException {
        if (!StringUtils.equalsIgnoreCase(replacementTable.getName(), expectedName)) {
            String message = "ReplacementTable API returned an entry we weren't asking for.";
            log.error(message);
            throw new NotCheckedException(message);
        }
    }

    private static void checkReplacementTableContent(ReplacementTable replacementTable, Map<String, String> expectedContent) throws NotCheckedException {
        for (Map.Entry<String, String> entry : expectedContent.entrySet()) {
            if (!replacementTable.getKeys().get(entry.getKey()).equals(entry.getValue())) {
                String message = "ReplacementTable with the name " + replacementTable.getName() + " exists, but does not equal to the given one.";
                throw new NotCheckedException(message);
            }
        }
    }

    public static boolean replacementTableExist(ReplacementTable replacementTable) throws NotCheckedException {
        ReplacementTable receivedReplacementTable = findReplacementTable(replacementTable.getName());

        if (receivedReplacementTable == null) {
            return false;
        }
        checkReplacementTableName(receivedReplacementTable, replacementTable.getName());
        checkReplacementTableContent(receivedReplacementTable, replacementTable.getKeys());
        return true;
    }

    public static boolean replacementTableExist(String replacementTableName) throws NotCheckedException {
        ReplacementTable receivedReplacementTable = findReplacementTable(replacementTableName);

        if (receivedReplacementTable == null) {
            return false;
        }
        checkReplacementTableName(receivedReplacementTable, replacementTableName);
        return true;
    }
```

**Comment by Arthasasyan** [comment]
Some points about how it work:

- Class `RequestSenderWithAuthorization` and it's subclasses are used to execute requests with Bearer or Basic authorization. All classes like `EntityMethod`, `MetricMethod`, etc. have a chain of methods, on the end of which an instance of `RequestSenderWithAuthorization` is taken and request is sent with the needed authorization. `RequestSenderWithBasicAuthorization.DEFAULT_BASIC_SENDER` is used to prevent creation of typical objects with authorization with credentials from config

- `DeletionCheck` is used to check if something was deleted. It takes any other `AbstractCheck` and checks if it did not pass

- `toParameterMap` method was added `MethodParameters` to convert it into format that can be queried. Some subclasses of `MethodParameters` have to override this method to set proper representation of their tags

**Comment by rodionos** [comment]
@Arthasasyan  - is this ready? needs review?

**Comment by Arthasasyan** [comment]
@rodionos this PR requires changes in many methods as authorization is modified a lot.
I believe that it can be really hard now to review all of it as there are 50 files changed. I suggest merging it if Travis CI checks are successfully passed and refactor in future PRs, because in other case there is a probability that this will never be merged at all
However, the final decision is on @raipc

**Pull request merged**